### ISSUE TITLE
kerning.plist is not sufficient to do mixed RTL and LTR kerning

### ISSUE THREAD

**Issue by graphicore** [unified-font-object/ufo-spec#16]
Without a hint to the kern feature generating code it is not possible to determine whether a kerning pair should be treated as LTR or RTL kerning. The kerning pair should be bear that information. External heuristics are probably bound to fail, but maybe a in most cases working solution is possible.

This was discovered in this discussion about Jomhuria, an Arabic script font that I work on:

https://groups.google.com/d/msg/googlefonts-discuss/kLnaUCegGjA/rjri18RMOvwJ

I quote here the relevant part:

> On 07/01/2015 01:01 PM, Khaled Hosny wrote:
> 
> > On Wed, Jul 01, 2015 at 05:35:50AM +0200, Lasse Fister wrote:
> > 
> > > On 07/01/2015 02:06 AM, Khaled Hosny wrote:
> > > 
> > > > On Tue, Jun 30, 2015 at 11:52:47PM +0200, Lasse Fister wrote:
> > > > 
> > > > > On 06/30/2015 11:19 PM, Khaled Hosny wrote:
> > > > > Also, where can I find the fea file to give it a look?
> > > > 
> > > > https://github.com/Tarobish/Jomhuria/blob/master/sources/kern-arabic.fea
> > > > That was created using ufo2fdk from a ufo created with GlyphsApp.
> > > 
> > > Hmm, I think the kerning is wrong here, I see things like:
> > > 
> > > pos @kern1.MMK_L_BehInti uni0676.fina -30;
> > > 
> > > But this will just subtract 30 units from the advance width of the rightmost
> > > glyph, which is not enough as you need to move it 30 units to the left
> > > as well, so it needs to be:
> > > 
> > > pos @kern1.MMK_L_BehInti uni0676.fina <-30 0 -30>;
> > 
> > Ok, so probably the ufo2fdk script should generate this for right to
> > left kerning, but it can't know from the way ufo stores its kerning
> > data. This is probably a problem with UFO as a format(?). UFO has only
> > very simple support for kerning:
> > 
> > http://unifiedfontobject.org/versions/ufo2/kerning.html
> > 
> > and also:
> > 
> > http://unifiedfontobject.org/versions/ufo3/kerning.html
> > 
> > The latter has a passage about writing direction:
> > 
> > > The kerning data is writing direction neutral. For text written
> > > left-to-right, the left-most glyph is the key in the top level
> > > dictionary. For text written right-to-left, the right-most glyph is
> > > the key in the top level dictionary. For example, given the pair /LG/,
> > > written left-to-right, the /L/ is the key in the top dictionary and
> > > the /G/ is the sub-dictionary. Given the pair /GL/, written
> > > right-to-left, the /G/ is the key in the top dictionary and the /L/ is
> > > the key in the sub-dictionary. 
> > 
> > But given that you say, that we need to know the writing direction to
> > make such a "simple" kern pair for RTL script, just stating that "data
> > is writing direction neutral" is not enough. Or do I miss here something?
> 
> I think so, there need to be a way to signal what direction the kerning
> will be used for (and it needs to be explicit, any heuristic is bounding
> to give wrong results for some cases).

My workaround at the moment is a subclass of `ufo2fdk.kernFeatureWriter` that writes only RTL kerning for all data of kerning.plist:

https://github.com/Tarobish/Jomhuria/blob/master/tools/getKernFeatureFromUFO.py

**Comment by behdad** [comment]
I don't agree that this needs to be in the API.  See: https://github.com/nodebox/opentype.js/issues/95#issuecomment-119471694 cc @khaledhosny

**Comment by graphicore** [comment]
@behdad from your linked comment

> Now, what this means for GPOS kerning is straightforward: if you are making changes to the left-side glyph, you update advance only. If you are changing the right-side glyph, you update both.

The point with this issue is that you can't know if you are changing the left side or the right side by looking at a UFO kerning.plist entry.

I agree that once you know which side you want to kern, there's no need for further arguments than one value. That's why it was easy to make a KernFeatureWriter for exclusively RTL kerning from ufo2fdk.

**Comment by benkiel** [comment]
I agree that pairs should have writing direction. However, I don't see a very clean way of doing so, without really breaking the current formatting of the kerning plist. My suggestion would be that the kerning value becomes a list, with the first member of the list the value, and the second the direction.

Perhaps this is a UFO4 item for consideration, however.

**Comment by benkiel** [comment]
@graphicore one thing I didn't think of this morning, assuming that your glyphs are all unicode encoded, it's very possible to pull direction from the python unicodedata.bidirectional property. Any kern pair with at least one member having a RTL property can be filtered for your kerning feature.

**Comment by graphicore** [comment]
First of all, I did not expect that the UFO v3 spec is open to change kerning.plist. I just report an issue that I have with it. I agree that UFO v4 could be a target and I'm also open for any solution.

I think, without fixing kerning.plist we have to use unicode and/or glyph names and class names.

Unicode can definitely be helpful here, but the unicode value of a glyph alone is not sufficient, because in Arabic you have at least four glyphs for one unicode-character. Three of these can be encoded (in Arabic Presentation Forms-A) but, afaik, for a modern environment you rather wouldn't do so because they are accessed via GSUB. The same is also true in Latin for stylistic sets and so on.

A naming scheme could help: One that let's you determine the unicode also for glyphs that are not encoded (like Adobes aglfn), plus maybe an optional "RTL" or "LTR" at the end of the glyph, that would be consulted first, if present. 

But, to be honest, that's not ideal. It's very indirect and it overloads the glyph names with a further use: writing features, pdf text extraction, referencing components and kerning directions. Not to forget finding glyphs in your editor and finding glyphs on the command-line when using git.

**Comment by graphicore** [comment]
> But, to be honest , that's not ideal.  …

Also, there are still cases then where you don't have full control. That may be corner cases—different writing directions are not really used within the same word, where kerning matters—but it's not a model that solves the full problem; so there might be bigger issues that we don't see yet.

**Comment by typesupply** [comment]
Yeah, I see the problem and it needs to be solved. To date, there have been RTL fonts produced with UFO, but the GPOS generation from the data in kerning.plist has had some additional smartness to organize the pairs by writing direction. ufo2fdk doesn't deal with this at all because the pair organization is very font specific.

One thing that I'd like to do is preserve the file structure of kerning.plist. The data structure in there is replicated all over the place and changing it could be problematic in the real world. I've thought of a couple of other ways to approach solving the problem:

1. Define some sort of pair organization algorithm or RTL glyph marking method. This will probably run into some nasty edge cases.
2. Add a new pair direction database somewhere in the UFO. For backwards compatibility, any pair defined in kerning.plist would be LTR unless defined as otherwise in the pair direction database. The potential problem with this is that a pair could only exist in one writing direction. I tried to think this through pretty thoroughly when I wrote the UFO 3 spec (I based it largely on the GPOS model) but I'm not a RTL expert by any means.
3. Add a new rtlkerning.plist file. This would keep them separate, wouldn't break anything that works now, but it would be a new file. I considered adding a vkerning.plist file in UFO 3 for a similar purpose, but I didn't get any feedback on the proposal so I scrapped it.

So, that's where the thinking on this is on my side of things.

**Comment by davelab6** [comment]
Lasse, in the short term, could you use 2 UFOs, and process one with all
the Arabic glyphs as entirely RTL, and the other with Lain as entirely LTR,
and then use pyftmerge to combine the binary results?

**Comment by graphicore** [comment]
Thanks Dave, I think that's what I'll do :-) But at the moment we have the Latin UFO that will have the Latin kerning, and a UFO that Kourosh produces with glyphs (from the TTF) where he kerns the Arabic. There are some cases like the parentheses from the Latin in Arabic where I am eager to see how it works out :-)

**Comment by LettError** [comment]
Are there use cases in which a single pair <firstglyph><secondglyph> would need different values for different directions?

**Comment by graphicore** [comment]
Not that I know, for a simple kerning pair. We need to know whether to produce:
LTR:
```
pos firstglyph secondglyph {value};
```
or RTL:
```
pos firstglyph secondglyph <{value} 0 {value} 0>;
```
But it's just one value.

@typesupply I think from your three proposals, the third proposal is the easiest to implement and the most robust of those three. The second one is worse than restructuring kerning data in UFO, especially because information that belongs together is stored at different places, so we spend our time syncing data then, it could make a good workaround however. The first one is also a possible workaround for the current situation.

**Comment by typesupply** [comment]
The possibility of <first><second> having different values when LTR and RTL seems illogical, but we know that crazy stuff exists (example: that very popular font that contains 2 glyphs with the same name; another one: those fonts that have two glyphs mapped to the same Unicode value). This could be tested by anyone with access to fonts containing both LTR and RTL kerning with a pretty simple fontTools script. It wouldn't be definitive, but it would be useful. I'll keep thinking about it.

I agree that redundant data is bad, but the nice thing about option 2 is that the data would be easily transferable between fonts. The data wouldn't have to be a 1:1 match to the contents of kerning.plist, it would be a general guide for separating the pairs. Still, it's a layer of complexity. UFO 4 is supposed to be only a structural change,  so I'm hesitant to introduce an entirely new file. But, it may be necessary.

I did have an idea on the existing implementation issue, specifically the RoboFab API. (kerning.plist grew out of the RoboFab API, which is a flat dict or pairs.) RoboFab's `__getitem__` and `get` methods could gain a new, optional, `direction` argument that defaults to `"ltr"`. Existing scripts would continue to work without any change.

**Comment by behdad** [comment]
Let me express my opinion again: the proposal here is wrong.

Kerning is an inherently visual procedure and is independent of the writing direction of the text.  The key point here is that kerning.plist specifically and rightly encodes "left glyph" and "right glyph".  So, for example, if someone is kerning the following sequence of two characters:
```
".
```
it absolutely doesn't matter if it's a double-quote followed by period, in a left-to-right script, or if it's period followed by double-quote in a right-to-left script.

Now.  The reason Lasse and possibly others are confused is how the GPOS table works.  Unfortunately individual lookups in OpenType do *not* specify what direction they expect their input in.  That's something we (OpenType stakeholders) are fully aware of and planning to fix, but in the mean time, the reality is that lookups in the GPOS table are processed in the *logical* text order, whatever that means.

For the most common cases, that's not a problem and "logical text order" is unambiguous.  For example, Latin is left-to-right, and Arabic is right-to-left.  That logic is encoded here: https://github.com/behdad/harfbuzz/blob/master/src/hb-common.cc#L446

The implication for this bug then is this: when generating a kern lookup, if it is to be referenced from a right-to-left script system ('arab' for example), it has to be generated differently from a lookup that is to be referenced from a left-to-right script system ('latn' for example).   You **cannot** use the same lookup.

Back to the `".` example, since both of those punctuation marks can happen in Arabic as well as Latin, you need to encode them in kern lookups referenced from both 'arab' and 'latn', and the encoding is different.  But as Erik very astutely asked, the value itself is not necessarily different and hence does not need to be reflected in the UFO file.

I'll go one step further with that: even if the desired kerning of the `".` sequence *was* different between Arabic and Latin, that's more of a script-specific or language-specific distinction, *not* direction-specific.  The same way that one might kern `".` differently in French vs English, or in Cyrillic vs Latin.  Again, nothing direction-specific.

Now, there's one case that is hugely ambiguous in OpenType: the case of kerning digits in Arabic and other right-to-left scripts, because digits in some of those scripts are written left-to-right.  But I don't want to get into that discussion here (and how inconsistent implementations are).

I hope I've made my argument clear.  Would be happy to clarify.

**Comment by graphicore** [comment]
@behdad thanks for your input. If I read you right, you still don't tell us how the current way the data is organized  in UFO makes it possible for us to determine the right output for our GPOS or AFDKO writing code.

> The implication for this bug then is this: when generating a kern lookup, if it is to be referenced from a right-to-left script system ('arab' for example), it has to be generated differently from a lookup that is to be referenced from a left-to-right script system ('latn' for example). You cannot use the same lookup.

Separating via script-language instead via script-direction sounds totally sane for me, however, *we still need that information to be present somewhere in UFO*.

Following the proposals we heard, the best bet is probably different kerning.plist files. But separated by language: `kerning-arab.plist`, `kerning-latn.plist` etc. That would serve me well. Also we can then set the language tag for the features and until OpenType is fixed produce RTL-specific positioning lookups for Arabic script.

> Now, there's one case that is hugely ambiguous in OpenType: the case of kerning digits in Arabic and other right-to-left scripts, because digits in some of those scripts are written left-to-right. But I don't want to get into that discussion here (and how inconsistent implementations are).

Thanks for pointing this out. But we need a solution for this problem here, we can't just not discuss it. You say we should use script language to gather the information needed to write the correct lookups, fine. But then you say this is broken when it comes to digits.
Should we use no language tag i.e. `kerning.plist` and there do the kerning of arabic numbers in a LTR fashion? Or is DFLT a better choice? Or do you imply we should implement the wisdom about those special conditions into the Arab feature writer? We could also just not kern these if it is to difficult to do in a consistent manner anyways i.e. FUBAR?

**Comment by behdad** [comment]
> @behdad thanks for your input. If I read you right, you still don't tell us how the current way the data is organized in UFO makes it possible for us to determine the right output for our GPOS or AFDKO writing code.

There's a lot of external data (from Unicode) that is needed to convert UFO to fonts, and this isn't any exception.  Here's the simplest heuristic that works:

  - Generate both an LTR kern lookup and an RTL one; including all kerning pairs in both.
  - Reference the RTL lookup from script systems that are RTL, and the LTR one for others.

That's the crudest heuristic, but it works.  From there on, you can add more complexity to achieve smaller kern tables.  But none of these really belongs into the UFO per se.  For example, a huge improvement:

  - Associate each glyph to a Unicode character,
  - Exclude from RTL kern table all glyphs associated with Unicode characters that have Bidi_Type=L,
  - Exclude from LTR kern table all glyphs associated with Unicode characters that have Bidi_Type=R or Bidi_Type=AL.


> Now, there's one case that is hugely ambiguous in OpenType: the case of kerning digits in Arabic and other right-to-left scripts, because digits in some of those scripts are written left-to-right. But I don't want to get into that discussion here (and how inconsistent implementations are).
> Thanks for pointing this out. But we need a solution for this problem here, we can't just not discuss it. You say we should use script language to gather the information needed to write the correct lookups, fine. But then you say this is broken when it comes to digits.
> Should we use no language tag i.e. kerning.plist and there do the kerning of arabic numbers in a LTR fashion? Or is DFLT a better choice? Or do you imply we should implement the wisdom about those special conditions into the Arab feature writer? We could also just not kern these if it is to difficult to do in a consistent manner anyways i.e. FUBAR?

In short, HarfBuzz and Uniscribe handle those differently.  Easiest is to ignore them for now.  Or if you wish, remove them from the kern pairs.  I can't recommend one way or another as the "right" way at this time.

**Comment by graphicore** [comment]
Great, thanks! I will try.

**Comment by twardoch** [comment]
I might add that there is also vertical kerning, in both directions. I agree that the most common understanding of "kerning" should be "one-dimensional positive or negative single-value adjustment in a given direction between two entities, each of them being a glyph or a group". Contextual or 2D adjustments should be done in FEA directly, but for "kerning pairs", it would be nice to have an ability to declare the directionality, which can be: 
- horizontal in any direction
- horizontal LTR
- horizontal RTL
- vertical in any direction
- vertical TTB
- vertical BTT

**Comment by davelab6** [comment]
> that very popular font that contains 2 glyphs with the same name

which font is that?

**Comment by benkiel** [comment]
Coming back to this, as we're thinking a bit more about UFO4, and it's something that's been on my mind.

I understand that @behdad is correct at a high level with this, but looking through the implementation of the feature writer, there is a lot to be argued for adding a direction tag (if not there, defaulting to ltr) to the kerning values. This would allow the designer to explicitly set pairs that they want written out in a direction, and avoid having to have complicated heuristics for unencoded glyphs just to write a kerning feature. It would also help kerning tools in displaying direction of a pair.

**Comment by behdad** [comment]
@benkiel What's the direction of the following pair: ")/"?

**Comment by behdad** [comment]
Also, there are plans to fix this in OpenType.  So, I suggest we think this through more before changing the spec to accommodate for currently font format shortcomings.

**Comment by benkiel** [comment]
@behdad I know, it's a mess. I'm glad to hear that it's being fixed in OpenType, but we may still need to think about this, as support for a new version of OpenType in existing engines will be slow.

If we add a tag, perhaps we define the following behavior for writing a kerning feature:
If the kerning has rtl, anything not tagged is written with both right to left and left to right lookups. Anything that is tagged ltr or rtl takes precedence, and is only written in the appropriate lookup.

This isn't as smart as your feature writing proposal, but it does have the advantage of capturing the designer's explicit intent.

**Comment by anthrotype** [comment]
> Kerning is an inherently visual procedure and is independent of the writing direction of the text.

I agree.

> The key point here is that kerning.plist specifically and rightly encodes "left glyph" and "right glyph".

That's not true, and that's probably the source of the confusion/misunderstandings about RTL kerning.

The "Writing Direction" paragraph from the [UFO3 kerning.plist spec](http://unifiedfontobject.org/versions/ufo3/kerning.plist/) says:

> The kerning data is writing direction neutral.

so far so good, it seems to agree with Behdad's comment above, i.e. the left is the left, the right is the right, no matter the script's writing direction. You have two shapes placed next to each other horizontally and you want to either reduce or increase the spacing between the two (by either changing the advance width of the left-hand-side glyph to move its right-sidebearing, or by changing both the advance width and the x-placement of the right-hand-side glyph to move its left-sidebearing).
However, the spec then continues on and (in my view) completely contradicts the previous point about writing direction neutrality:

> For text written left-to-right, the left-most glyph is the key in the top level dictionary. For text written right-to-left, the right-most glyph is the key in the top level dictionary.

Now, if kerning were neutral and visual, the key in the top-level dictionary would _not_ change depending on the writing direction, i.e. it would always be the _left_ glyph or group.
Whereas according to the cited passage from the UFO3 spec, the key glyph/group in the kerning dictionary depends on the direction of the text (although this direction is not specified anywhere and _that_ is the problem Lasse brought up). The terminology used for the kerning groups (first and second, instead of left and right as in the old MMK_ classes) also reinforces this point. Basically, it seems that the UFO kerning.plist follows the same "logical" text order in which OpenType lookups are processed.

The paragraph ends with an example which can be confusing at first, because the RTL case is not a real-world one (using for example two Arabic letters)

> For example, given the pair LG, written left-to-right, the L is the key in the top dictionary and the G is the sub-dictionary. Given the pair GL, written right-to-left, the G is the key in the top dictionary and the L is the key in the sub-dictionary.

So... we need to revive this discussion and agreee on a fix. We either add some pair direction database to mark each pair as LTR only, RTL only or both, or we change the semantic of kerning.plist to truly be visual left and right.

**Comment by anthrotype** [comment]
cc @behdad

**Comment by LettError** [comment]
> On 30 Jan 2018, at 22:01, Cosimo Lupo <notifications@github.com> wrote:
> 
> so far so good, it seems to agree with Behdad's comment above, i.e. the left is the left, the right is the right, no matter the script's writing direction. You have two shapes placed next to each other horizontally and you want to either reduce or increase the spacing between the two (by either changing the advance width of the left-hand-side glyph to move its right-sidebearing, or by changing both the advance width and the x-placement of the right-hand-side glyph to move its left-sidebearing).

What about vertical?

E

**Comment by benkiel** [comment]
LTF, RTL, Horizontal (both), Vertical?

**Comment by typesupply** [comment]
> Basically, it seems that the UFO kerning.plist follows the same "logical" text order in which OpenType lookups are processed.

Right, I modeled it on the GPOS spec. However, as @behdad notes, the GPOS spec was confusing in itself. To be perfectly honest, I never felt like I had a firm grasp on this when I wrote the spec. There wasn't much documentation available at the time and I didn't have any significant experts to consult. So, if it's wrong or unclear, I'm 100% in favor of removing my RTL notes in favor of something better. I don't think there is a terribly important backwards compatibility concern or unresolvable legacy data on this that we need to worry about.

> So... we need to revive this discussion and agreee on a fix. We either add some pair direction database to mark each pair as LTR only, RTL only or both, or we change the semantic of kerning.plist to truly be visual left and right.

My tiny bit of caution here: as UFO 3 has demonstrated, significant changes to existing storage structures are very difficult to get into implementations. Thus, I favor leaving `kerning.plist` as is and adding one or more supplemental files. What those are...I don't know since I know no more about RTL than I did before. While we are at this, we should handle vertical kerning as well.

Any proposals?

**Comment by LettError** [comment]
+1 for not breaking `kerning.plist`. We could add `gninrek.plist` for RTL.

**Comment by moyogo** [comment]
How about keeping kerning.plist for LTR, adding kerningrtl.plist for RTL
and kerningttb.plist for top-to-bottom?

**Comment by benkiel** [comment]
Notes on this from Robothon: We agreed in principal to add a vkerning.plist and a hkerning.plist. These values will be specified as a pair being visual left and visual right to clean up the first/second writing direction ambiguity. If a UFO has hkerning.plist, it must not have kerning.plist. We also discussed changing the kerning value to be a list, where length of list must be one less than the length of the pairs. This allows for contextual kerning. A sample algorithm for writing out GPOS dealing with RTL/LTR and contextual kerning would be provided.

**Comment by behdad** [comment]
> We also discussed changing the kerning value to be a list, where length of list must be one less than the length of the pairs. This allows for contextual kerning.

Can you elaborate on this? It's trickier than that to specify what part constitutes context and what match.

**Comment by behdad** [comment]
> We also discussed changing the kerning value to be a list, where length of list must be one less than the length of the pairs. This allows for contextual kerning.
> 
> Can you elaborate on this? It's trickier than that to specify what part constitutes context and what match.

Or rather. That should be handled in a separate issue.

**Comment by benkiel** [comment]
@behdad will do tomorrow, but this is the time to handle contextual kerning, doing it later will just be another breaking change, better to do this once. Will open an issue tomorrow on it.

**Comment by justvanrossum** [comment]
Hi, 2019 here. I assume people are still using custom kern writers to create RTL GPOS kerning? What is the most common solution today in fontmake land?

How can UFO make it easier for people to include RTL kerning in their fonts?

In #87, @typoman suggests to add a direction property to glyphs, but I'm not convinced this solves much unless compiler writers find it an appealing solution.

**Comment by schriftgestalt** [comment]
I didn’t read the full thread (yet) but in Glyphs the kern feature is written from a single kerning structure and outputs correct LTR and RTL lookups. So it is possible if all glyph info is set properly.

**Comment by justvanrossum** [comment]
> I didn’t read the full thread (yet) but in Glyphs the kern feature is written from a single kerning structure and outputs correct LTR and RTL lookups. So it is possible if all glyph info is set properly.

Do you use a similar scheme as what Behdad suggested? https://github.com/unified-font-object/ufo-spec/issues/16#issuecomment-120036174

**Comment by schriftgestalt** [comment]
I use something similar to the second approach. With hard coded exceptions for numbers and bidi glyphs.

**Comment by typoman** [comment]
I don't think there is a reason this issue is open as ~~the kerning value records in GPOS can be separated by script not the direction of the glyph.~~ #88

**Comment by typoman** [comment]
> kerning value records in GPOS can be separated by script not the direction of the glyph.

Incorrect: https://github.com/unified-font-object/ufo-spec/issues/88#issuecomment-488694343

**Comment by anthrotype** [comment]
> I assume people are still using custom kern writers to create RTL GPOS kerning? What is the most
common solution today in fontmake land?

I'll try to explain how the ufo2ft KernFeatureWriter used by fontmake currently deals with mixed-direction kerning. **TLDR: yeah, it's complicated**.

The algorithm for generating  FEA kern feature from UFO kerning is the following:

- parse GDEF GlyphClassDefinition from UFO features.fea to get the set of "Mark" glyphs (this will be used later to decide whether to add `ignoreMarks` flag to kern lookups containing pairs between base and mark glyphs).
- get `public.kerning{1,2}` classes from groups.plist
- get kerning pairs from kerning.plist
- parse `languagesystem` statements (script/language pairs) in features.fea, and group them by the global horizontal direction of the script, and by whether the given script requires the 'dist' feature (Indic) or it can use 'kern'.
- build a unicode to glyph name mapping (cmap) from each UFO glif's unicode value
- When there are any characters from globally RTL scripts in the cmap, we take the UFO features.fea so far and compile a GSUB table from it. Then we use fonttools subsetter to follow the glyph substitutions from cmapped glyphs: i.e. given an initial set of glyphs that have a unicode assigned, we perform a 'closure' over the GSUB table (the same way the subsetter does to decide whether to keep "f_i" when you request to keep characters "U+0066,U+0069" plus the "liga" feature). With this we obtain the set of all glyphs that are reachable via substitutions from this intial set. We use it to classify all glyphs in the font by the following unicode properties: the global horizontal direction of the script that a given unicode character belongs, and its bidirectional type (See `classifyGlyphs` function in `ufo2ft.util`).
- We mark each kerning pair with these script directions and bidiTypes when _any_ of the glyphs involved in the kern pair (not only the first) intersects one of these glyph classes (Of course since multiple unicodes can be mapped to the same glyph, these classifications may be ambiguous... let's ignore for now). 
- To get the script direction for a unicode character, we use `fontTools.unicodedata.script` to obtain the script code from the unicode codepoint, and then `fontTools.unicodedata.script_horizontal_direction` (which uses an hard-coded list of RTL scripts from harfbuzz source code and CLDR).
- If there is no RTL-script glyph in all kerning pairs, we create a single (implicitly LTR) kern lookup and register it under all (non-RTL) scripts, and we go home.
- If there are any RTL-script glyphs among the kerning pairs, we split kerning into two or three lookups: one DFLT lookup for 'Common' (`Zyyy`) script-agnostic glyphs (like digits or punctuation) and two LTR and RTL lookups.
- We drop all the kerning pairs whose glyphs have an ambiguous script direction (both LTR and RTL), as they won't work (issue a warning).
- By default KernFeatureWriter adds the ignoreMarks lookupflag to all generated lookups (but can be globally disabled). However, when kerning pairs contain any glyphs that are classified as "Mark" in GDEF (as may happen in some Indic fonts, like Noto Sans Siddham), the writer splits each lookup into two: one with ignoreMarks flag containing base-to-base pairs, and one without ignoreMarks containing base/mark pairs (or even mark/mark).
- We now have this list of all valid kerning pairs, each associated with a global script direction (either LTR, RTL or None), and with a set of bidiTypes. The latter are simplified into four classes, idendified by a symbolic constant:
  1) "L" (for "strong" left-to-right bidi type),
  2) "R" (for "strong" right-to-left bidi type, including both "R" and "AL" [arabic letter] bidi types);
  3) "N" for numbers which have a "weak" left-to-right direction (includes both "EN" [english/european? numbers] and "AN" [arabic numbers] bidi categories);
  4) and 'None' if it is neither of the above classes (i.e. "neutral" bidirectional category)
- From these we create three lookups. The first 'kern_dflt' lookup excludes all pairs with either LTR or RTL script direction; only includes pairs whose script direction is None (e.g. Hebrew is RTL but uses latin digits, which are in 'Common' script and have "weak" left-to-right bidi type); this kern_dflt lookup is registered under all the languagesystems. Since we don't know what the original intention of the designer was, we assume that these pairs are kerned left-to-right (i.e. only modify the xAdvance of the first glyph). There is no logical 'first' and 'second' when glyphs have no script (i.e. may occur in either LTR or RTL script) and have neutral bidi type. So we assume that by 'first' the designer means 'left' (it sucks.. This could be fixed by having that new hkerning.plist where glyphs are kerned in visual order, left and right, not logical first and second.. but let me finish this first).
- The second is 'kern_ltr' lookup which excludes all pairs that either have no script direction (already in kern_dflt) or RTL script direction, only includes LTR-script pairs. We register this lookup under the LTR languagesystems. The kern pairs will be encoded with a simple LTR valuerecord that only modifies the advance width of first glyph.
- Finally we make a 'kern_rtl' lookup which excludes pairs with no script direction or LTR-script, includes only RTL-script pairs, and is registered under the RTL languagesystems. We normally encode these RTL pairs using the explicit four-value notation (`<-80 0 -80 0>`). That's because, in order to move the left sidebearing of the first glyph (the _right_-most glyph), one needs to modify both its xPlacement and xAdvance. We do that for all pairs in kern_rtl -- **except** the pairs that contain glyphs with "N" bidi type (i.e. arabic numbers) which should be encoded as left-to-right even though they belong to the RTL kern lookup (for RTL scripts).
- Finally, for dist-enabled Indic scripts, we do _not_ split the pairs associated with "Indic" scripts in separate lookups, one under dist, the rest under kern, but we simply register the same 'kern_ltr' and/or 'kern_rtl' lookups (as above) under the "dist" feature instead of "kern" (this is simpler like this and works just fine).

I'll let this sink in a bit.. :sweat_smile:

**Comment by benkiel** [comment]
@anthrotype. Woah. Thank you for that, it helps make things clearer and I know it took a long time to write. Beyond hkerning.plist is there anything that could be added to the UFO that would be useful here?

I'm also curious to know what @TiroTypeworks does for kern feature generation

**Comment by justvanrossum** [comment]
Yes, thanks @anthrotype, for this super detailed description. Very useful and informative.

**Comment by typoman** [comment]
Thank you @anthrotype. I think parsing GSUB part that you explained could be the ultimate answer to all the problems I was trying to address.

**Comment by schriftgestalt** [comment]
Thanks for the explanation. 

I would not use "visual order" but always writing order. That is as euro centric as it gets. And it doesn't make the generation of the kern feature any simpler. Just call it `kerningRTL` and store the pairs in text order. Maybe call them firstPair and secondPair instead of left and right.

**Comment by typoman** [comment]
> I would not use "visual order" but always writing order.
> Maybe call them firstPair and secondPair instead of left and right.

Same here for both. It's very confusing when any entry from pairs are called left or right.

**Comment by tiroj** [comment]
> I'm also curious to know what @TiroTypeworks does for kern feature generation

Historically, we've created separate FontLab Studio 5 kern sources for LTR and RTL kerning, and have scripts that write VOLT .vtl lookup (and .vtg class) files, which then get imported in our our VOLT projects. It's a solid workflow that we've used on many complex projects for a long time.

I'm still trying to figure out the best way forward with new tools, especially in variable fonts. The [Volto](https://github.com/TiroTypeworks/Volto) project that we published recently is part of that process: a way of getting OTL data from VOLT into .fea code which we can then experiment with in different workflows. But so far I've not tried dealing with RTL kerning in any of these scenarios.

VOLT remains my preferred OTL tool, so my real hope is that MS open sources it and someone rewrites it in Python or, frankly, any language other than Visual Basic, with variable font support.

**Comment by tiroj** [comment]
> Maybe call them firstPair and secondPair instead of left and right.

Absolutely. You have to think of kerning in terms of first and second glyphs/glasses, and not left and right. Ideally, one should have the option to flip the entire kerning UI when doing RTL kerning, so that it is possible to work visually with first=right and second=left if that's your preference.

**Comment by typoman** [comment]
> so my real hope is that MS open sources it and someone rewrites it in Python or, frankly, any language other than Visual Basic, with variable font support.

I would like to take this opportunity to ask @PeterCon to consider this or at least a compiler that is platform independent. VOLT could be used by type designers who don't have Windows and I think they're a lot.

**Comment by benkiel** [comment]
@tiroj Thank you for chiming in; it's helpful to get your perspective.

> Absolutely. You have to think of kerning in terms of first and second glyphs/glasses, and not left and right.

It has taken me a bit to wrap my head around what @behdad says above (and in person at Robothon): which is that in some cases you don't know what is first/second when there is no writing direction attached to a glyph (say the pair period/parenright). The kerning for that pair would be very different if written LTR vs RTL, so being able to encode in your kerning which glyph is on which side when you made spacial adjustment is important, which is where left/right comes in. 

Yes, it gets confusing, but what the UFO needs to do is to encode in the kerning what space adjustment was desired, and give enough to a complier to be able to make a feature. For mixed writing directions, this will always be complex, but it would be nice to improve the storage to make a bit easier.

**Comment by schriftgestalt** [comment]
@benkiel @behdad  I would like to see an example.

**Comment by behdad** [comment]
> @benkiel @behdad I would like to see an example.

Example of what?  A kern pair that doesn't have a "clear" direction? period/parenright

**Comment by schriftgestalt** [comment]
Yes.

**Comment by tiroj** [comment]
> A kern pair that doesn't have a "clear" direction? period/parenright

The direction has to be determinable first from the text, based on the BiDi algorithm, and the punctuation rolled into glyph runs of appropriate directionality with adjacent glyphs, before we can even begin to address the question of how to make the lookup directionality clear. Do you think that's the case, Behdad: can we reliably determine what the direction of punctuation characters for OTL processing should be from BiDi (taking into account the complexities of embedding levels)?

**Comment by behdad** [comment]
> > A kern pair that doesn't have a "clear" direction? period/parenright
> 
> The direction has to be determinable first from the text, based on the BiDi algorithm, and the punctuation rolled into glyph runs of appropriate directionality with adjacent glyphs, before we can even begin to address the question of how to make the lookup directionality clear. Do you think that's the case, Behdad: can we reliably determine what the direction of punctuation characters for OTL processing should be from BiDi (taking into account the complexities of embedding levels)?

Yes. But also means that the kern pair for punctuations should be encoded in all script systems in the font.

**Comment by anthrotype** [comment]
That's the problem. If kerning.plist contains such a pair between punctuation, e.g. `(period, parenleft): -30`, if one would like to encode this pair in both the kern lookup for LTR scripts and in the one for the RTL scripts, one needs to know (or make assumptions as to) what the original intention of the font designer was when kerning these two direction-neutral glyphs. If the UFO contained a new hkerning.plist (and vkerning.plist) which stored the kerning pairs not as first/second logical order (which is script-writing-direction depdendent), but as left and right glyph (or top and bottom, no matter what the text direction), then the compiler could write the same kerning pairs in both LTR and RTL.
E.g. let's pretend `(period, parenleft): -30` in the example is meant to be kerned with period glyph on the left and parenleft on its right. Then for the LTR lookup, the generated rule would be `pos period parenleft -30` (i.e. reduce period's advance width by 30 units); whereas in the RTL lookup it should be `pos parenleft period <-30 0 -30 0>` (move the left sidebearing of parenleft by reducing both its xAdvance and xPlacement by 30 units).

**Comment by tiroj** [comment]
> Yes. But also means that the kern pair for punctuations should be encoded in all script systems in the font.

Agreed: that's standard practice here. When making multi-script fonts, I've tended to create separate kern sources for each script, so our FL-to-VOLT workflow can easily generate separate sets of lookups for each script, including punctuation kerning, numeral kerning, etc.. Of course, there's some punctuation characters that are never used with some scripts, so I also tend to remove kerning pairs for those characters from the sources for those scripts.

I don't know a robust way to automate splitting of kerning lookups by script, which is why I favour separate sources that I manage myself. One wants a kerning tool to be able to write the OTL lookups from whatever native format the tool uses, and to do so efficiently in terms of both data and layout processing, and to my mind that means controlling the separation of GPOS lookups that receive the kerning for different scripts. Since those lookups also need to include common punctuation glyphs — for kerning to the script and to each other, possibly in multiple directions —, keeping the sources separate has so far been the safest approach.

**Comment by typoman** [comment]
> Then for the LTR lookup, the generated rule would be `pos period parenleft -30` (i.e. reduce period's advance width by 30 units); whereas in the RTL lookup it should be `pos parenleft period <-30 0 -30 0>` (move the left sidebearing of parenleft by reducing both its xAdvance and xPlacement by 30 units).

Does that mean Arabic numbers kerning which is also written in the RTL lookup should have the `<-30 0 -30 0>` format?

**Comment by behdad** [comment]
> That's the problem. If kerning.plist contains such a pair between punctuation, e.g. `(period, parenleft): -30`, if one would like to encode this pair in both the kern lookup for LTR scripts and in the one for the RTL scripts, one needs to know (or make assumptions as to) what the original intention of the font designer was when kerning these two direction-neutral glyphs.

Right.

UFO2 was fine technically, by using "left" and "right" words.  UFO3 departed from this in the name of being direction-agnostic, by calling the glyphs in the pair "first" and "second", in practice making it inherently ambiguous.

**Comment by anthrotype** [comment]
> Does that mean Arabic numbers kerning which is also written in the RTL lookup should have the <-30 0 -30 0> format?

no, that's the exception to the rule. They have "AN" bidirectional category which is weak LTR.

**Comment by tiroj** [comment]
@behdad I recall a conversation at the OTWG meeting in Mountain View in which we discussed the impact for kerning of Arab/Persian numerals in Harfbuzz's handling of Arabic layout vs Uniscribe's. My recollection is that if a font contains LTR kerning for numerals, presuming Uniscribe's treatment of numbers as LTR runs, this could be misinterpreted in Harfbuzz because of the way you apply RTL directionality across Arabic layout? Is that right?

**Comment by anthrotype** [comment]
to clarify once again, Arabic number kerning should be written in the kern lookup for the RTL scripts, but using a LTR pos rule (only modifying advance of first glyph).

**Comment by justvanrossum** [comment]
> UFO2 was fine technically, by using "left" and "right" words. UFO3 departed from this in the name of being direction-agnostic, by calling the glyphs in the pair "first" and "second", in practice making it inherently ambiguous.

Yes, that was a mistake, and I think our challenge is to come up with a clean way to correct this. Adding a new `hkerning.plist` file would be a good solution. We're exploring some alternatives, though.

**Comment by behdad** [comment]
> @behdad I recall a conversation at the OTWG meeting in Mountain View in which we discussed the impact for kerning of Arab/Persian numerals in Harfbuzz's handling of Arabic layout vs Uniscribe's. My recollection is that if a font contains LTR kerning for numerals, presuming Uniscribe's treatment of numbers as LTR runs, this could be misinterpreted in Harfbuzz because of the way you apply RTL directionality across Arabic layout? Is that right?

Correct.  HarfBuzz applies kerning in RTL to all script=Arabic text.  Uniscribe on the other hand, uses a hack, by classifying Arabic numerals as a different ArabicNumeral script code, which has direction LTR.

> to clarify once again, Arabic number kerning should be written in the kern lookup for the RTL scripts, but using a LTR pos rule (only modifying advance of first glyph).

That's what Uniscribe does.  I wouldn't call it a rule though.  It's a hack we have to live with.

**Comment by tiroj** [comment]
If glyphs are laid out left-to-right, it makes sense to me that kerning is applied left-to-right too. Applying kerning right-to-left for Arab/Persian numerals is really confusing, speaking as a font developer.

**Comment by behdad** [comment]
> If glyphs are laid out left-to-right, it makes sense to me that kerning is applied left-to-right too. Applying kerning right-to-left for Arab/Persian numerals is really confusing, speaking as a font developer.

(Why did you have to open that can of worms here...)

Right.  But speaking as a shaper, when I see script=Arab direction=LTR, I don't know if it's numbers running their natural LTR, or letters forced to LTR visa LRO character.  Short of looking at the character codepoints I have no way to distinguish.

**Comment by typoman** [comment]
@behdad I'm a bit puzzled. Could you please elaborate why direction agnostic kerning pairs in the new UFO format is ambiguous if at the end all the pairs should be written in a logical order?

**Comment by justvanrossum** [comment]
> I'm a bit puzzled. Could you please elaborate why direction agnostic kerning pairs in the new UFO format is ambiguous if at the end all the pairs should be written in a logical order?

I'm not Behdad, but this is how I understand it: take a "neutral" kern pair, such as period/parenright. If the order in the kerning data is visual left-to-right, it is clear what is meant if you kern that pair, and you can compile it to LTR kerning and RTL kerning.

If it's supposed to follow the writing direction (as in the UFO3 spec), it's impossible to determine what the _intended_ direction of the the pair was, as the glyphs in question have no inherent direction. So you simply can't tell whether "period parenright 40" was meant as visual "period parenright 40" or visual "parenright period 40". If the intention is not clear, how can you even compile it?

This seems to be confirmed by this quote from Cosimo's explanation:

> There is no logical 'first' and 'second' when glyphs have no script (i.e. may occur in either LTR or RTL script) and have neutral bidi type. So we assume that by 'first' the designer means 'left' (it sucks.. This could be fixed by having that new hkerning.plist where glyphs are kerned in visual order, left and right, not logical first and second.

**Comment by anthrotype** [comment]
correct. the same thing I wrote earlier in this thread:
https://github.com/unified-font-object/ufo-spec/issues/16#issuecomment-489180090

**Comment by typoman** [comment]
so hypothetically if the value in the pair was written in four int array but still in a logical order it would be fine?

**Comment by justvanrossum** [comment]
> so hypothetically if the value in the pair was written in four int array but still in a logical order it would be fine?

You need to separate what is (or can be/should be) stored in the UFO and what has to be compiled to GPOS. Those are two different things. The fact that RTL kerning is "difficult" in GPOS should have no influence on the UFO spec.

**Comment by typoman** [comment]
I'm only trying to figure out how a kerning tool should store the data and how should it compile it in UFO. I'm very sorry if I get things slow, but since I've only understood the direction agnostic way of storing data, the idea of storing pairs in visual order is throwing me out of balance. Maybe examples will make it easy to implement:

UFO3 (kerning pair direction agnostic):
- UFO pair `period parenright 40`
- GPOS value record:
	RTL scripts: `period parenright <40 0 40 0>`
	LTR scripts: `period parenright 40`

UFO2 (kerning pair from left to right):
- UFO pair `period parenright 40`
- GPOS value record:
	RTL scripts: `parenright period <40 0 40 0>`
	LTR scripts: `period parenright 40`

Maybe the above example is wrongly written in the GPOS record. But If I'm correct, how UFO2 is a better approach?
Edit: made a correction on UFO3 example

**Comment by justvanrossum** [comment]
Let's take this off line. Please email me.

**Comment by anthrotype** [comment]
UFO3 kerning is _not_ writing direction agnostic, it uses logical (first/second) order like in OpenType, thus depends on the script writing direction what is deemed "first" or "second", hence not writing direction agnostic.
UFO2 kerning was writing direction agnostic in that left and right are always left and right no matter from which perpective you look at it; it's not implying left is first, or that second is right -- to the contrary. I called it "visual" order but that may the source of the confusion. Call it maybe "natural" if you wish, as opposed to "logical". It simply means that two glyphs are placed next to each other and will be spaced more tightly or loosely when they occur next to each other *no matter whetehr one comes first and the other second or viceversa*

**Comment by anthrotype** [comment]
if you believe that storing the left-most glyph in a pair on the left-hand side, and the right-most glyph on the right-hand side is enthocentrism... well, I'm ok to swap the order if that makes it more comfortable for you, I don't really care :)

**Comment by typesupply** [comment]
I'm going to close this as we have a discussion about how to fix this in #96.
